### PR TITLE
Implement Non Fungible Token Royalty (EIP2981)

### LINKED ISSUE #2789: Implementing EIP-2981 (ERC721 Royalties)
**🧐 Motivation**

Recently, [EIP-2981](https://eips.ethereum.org/EIPS/eip-2981) went to Final Status. It proposes a "standardized way to retrieve royalty payment information for non-fungible tokens (NFTs) to enable universal support for royalty payments across all NFT marketplaces and ecosystem participants".

While it is possible to implement your own royalty system overriding ERC721/ERC1155's `transferFrom` functions, I was wondering if it would make sense to support EIP-2981 as an alternative to minimize gas usage in deployment.  Plus, EIP-2981 would better support [optional royalty payments](https://eips.ethereum.org/EIPS/eip-2981#optional-royalty-payments).

**📝 Details**

If it makes sense to add EIP-2981 support to OpenZeppelin contracts, I think it could be a nice idea to make an interface ERC721HasRoyalties.sol and IERC1155HasRoyalties.sol to extend ERC721 and ERC1155 contracts.

(P.S.: Since it has been recently finalized, I don't think many marketplaces are adopting EIP-2981 for now, but if this feature makes sense, I can keep track of it 😃)

### PR DESCRIPTION
**Pull request by JulissaDantes**
Fixes #2789  

Implementation of NFT Royalties according to the [EIP-2981](https://eips.ethereum.org/EIPS/eip-2981). This implemementation supports both global royalties and token based royalties. 

#### PR Checklist

- [x] Tests
- [x] Documentation
- [x] Changelog entry

### CODE REVIEW

File: contracts/token/ERC721/extensions/draft-ERC721Royalty.sol
Comment (frangio) [inline contracts/token/ERC721/extensions/draft-ERC721Royalty.sol:36]: There is no way to delete token royalty if it needs to be reset back to global royalty. We might need a separate function to do that.

File: contracts/token/ERC721/extensions/draft-ERC721Royalty.sol
Comment (frangio) [inline contracts/token/ERC721/extensions/draft-ERC721Royalty.sol:69]: We generally avoid this mechanism of defining return variables, and instead prefer using explicit `return` statements.

File: test/token/ERC721/extensions/ERC721Royalty.test.js
Comment (frangio) [inline test/token/ERC721/extensions/ERC721Royalty.test.js:62]: Does the EIP disallow a 100% royalty amount? It's a bit ridiculous maybe but I think it should be allowed. Unless the EIP says otherwise, I personally think the royalty percentage could be set to >100%... HOWEVER, as a sanity check and general protection I think our implementation should not allow >100%.

File: contracts/token/ERC721/extensions/draft-ERC721Royalty.sol
Comment (Amxx) [inline contracts/token/ERC721/extensions/draft-ERC721Royalty.sol:8]: this is nice for testing, but should be removed before merging.

File: contracts/token/ERC721/extensions/draft-ERC721Royalty.sol
Comment (Amxx) [inline contracts/token/ERC721/extensions/draft-ERC721Royalty.sol:80]: IMO, we need an `_burn` override that deletes the royaltyInfo if a token is burned (similarly to what ERC721UriStorage does)

File: contracts/token/ERC721/extensions/draft-ERC721Royalty.sol
Comment (JulissaDantes) [reply contracts/token/ERC721/extensions/draft-ERC721Royalty.sol:39]: Since this is an extension for a ERC721 and not an `ERC721` we cannot check if it was minted, but its on the documentation requirements that it should be mined already

File: contracts/token/ERC721/extensions/draft-ERC721Royalty.sol
Comment (JulissaDantes) [reply contracts/token/ERC721/extensions/draft-ERC721Royalty.sol:113]: Since this doesnt inherits from `ERC721` I cannot override burn, but I'll include a method to delete a token royalty and include in the documentation that it should be called when burning

File: contracts/token/ERC721/extensions/draft-ERC721Royalty.sol
Comment (JulissaDantes) [reply contracts/token/ERC721/extensions/draft-ERC721Royalty.sol:40]: yes

File: contracts/token/ERC721/extensions/draft-ERC721Royalty.sol
Comment (Amxx) [inline contracts/token/ERC721/extensions/draft-ERC721Royalty.sol:108]: Do we really need the if here ?

File: contracts/interfaces/IERC2981.sol
Comment (frangio) [inline contracts/interfaces/IERC2981.sol:22]: I believe this is actually `royaltyAmount`, right? The fraction is the %, but this should be amount based on the sale price.

File: contracts/token/ERC721/extensions/draft-IERC721Royalty.sol
Comment (frangio) [inline contracts/token/ERC721/extensions/draft-IERC721Royalty.sol:28]: `royaltyAmount` here as well.

File: contracts/token/ERC721/extensions/draft-IERC721Royalty.sol
Comment (frangio) [inline contracts/token/ERC721/extensions/draft-IERC721Royalty.sol:13]: This just inserts an extra newline for formatting:

```suggestion
 *
 * _Available since v4.5._
```

File: contracts/token/ERC721/extensions/draft-ERC721Royalty.sol
Comment (frangio) [inline contracts/token/ERC721/extensions/draft-ERC721Royalty.sol:41]: I think we need to allow 0. It would be the only workaround to have royalties globally but disable them per-token.

**Review by frangio — COMMENTED** [review]
This is looking good!

File: contracts/token/ERC721/extensions/draft-ERC721Royalty.sol
Comment (JulissaDantes) [reply contracts/token/ERC721/extensions/draft-ERC721Royalty.sol:41]: But you could call the deleteTokenRoyalty function

File: contracts/interfaces/IERC2981.sol
Comment (JulissaDantes) [reply contracts/interfaces/IERC2981.sol:22]: Yes

File: contracts/token/ERC721/extensions/draft-ERC721Royalty.sol
Comment (Amxx) [inline contracts/token/ERC721/extensions/draft-ERC721Royalty.sol:41]: I think this function could be used to set but also to update ... and updating in a way that sets the fee to 0 might be acceptable.

Also, if the global fee is not 0, maybe setting a custom token fee to 0 makes sens.

So I would remove that check.

File: contracts/token/ERC721/extensions/draft-ERC721Royalty.sol
Comment (Amxx) [inline contracts/token/ERC721/extensions/draft-ERC721Royalty.sol:59]: same argument: this function might be used to "update" ... and putting 0 could make sens

File: contracts/token/ERC721/extensions/draft-ERC721Royalty.sol
Comment (Amxx) [inline contracts/token/ERC721/extensions/draft-ERC721Royalty.sol:22]: if the denominator is 10000, then we could possibly use a uint96 so that the info fit into a single slot

File: contracts/token/ERC721/extensions/draft-ERC721Royalty.sol
Comment (Amxx) [inline contracts/token/ERC721/extensions/draft-ERC721Royalty.sol:72]: if we change the `royaltyFraction` to be a uint96, then `RoyaltyInfo` would be a single slot ... which could then be optimized to avoid doing the ssload twice:

```
RoyaltyInfo memory royalty = _tokenRoyaltyInfo[_tokenId];
if (royalty.receiver == address(0)) {
    royalty = _globalRoyaltyInfo;
}

...
```

File: contracts/token/ERC721/extensions/draft-ERC721Royalty.sol
Comment (Amxx) [inline contracts/token/ERC721/extensions/draft-ERC721Royalty.sol:86]: I would personnaly put that at the top, so its not lost in the middle of royalty-related code

File: contracts/token/ERC721/extensions/draft-ERC721Royalty.sol
Comment (Amxx) [inline contracts/token/ERC721/extensions/draft-ERC721Royalty.sol:94]: possibly change that to `uint96`

File: contracts/token/ERC721/extensions/draft-ERC721Royalty.sol
Comment (Amxx) [inline contracts/token/ERC721/extensions/draft-ERC721Royalty.sol:107]: there is no version of that for the global ?

File: contracts/token/ERC721/extensions/draft-ERC721Royalty.sol
Comment (JulissaDantes) [reply contracts/token/ERC721/extensions/draft-ERC721Royalty.sol:41]: Yes, I'll remove it, and also they can use to update the royalty not just set it.

File: contracts/token/ERC721/extensions/draft-ERC721Royalty.sol
Comment (JulissaDantes) [reply contracts/token/ERC721/extensions/draft-ERC721Royalty.sol:107]: You are right, is missing, I'll add it

Comment (Amxx) [reply contracts/token/ERC721/extensions/draft-ERC721Royalty.sol:107]: Note that if you remove the checks `fraction > 0` and `recipient != address(0)` this could possibly not be needed

maybe these two checks can be merged into 
```
require(recipient != address(0) || fraction == 0);
```

Comment (Amxx) [reply contracts/token/ERC721/extensions/draft-ERC721Royalty.sol:102]: Wait upon further thinking it should be or and not xor.

We want to let token have a royalty fraction of 0 with a recipient that is not 0 → this is how you enforce that a token has fraction 0 when the global is not 0

So
```
require(recipient != address(0) || fraction == 0);
```

Comment (JulissaDantes) [reply contracts/token/ERC721/extensions/draft-ERC721Royalty.sol:102]: I've also included a new test case for this scenario

**Review by Amxx — APPROVED** [review]

**Review by frangio — COMMENTED** [review]
I noticed two things: 1) the ERC is already Final, so we can remove the `draft-` prefix from files, 2) we already had the interface in `interfaces`, so we should remove `IERC721Royalty` and use `IERC2981` that was already there.

File: test/token/ERC721/extensions/ERC721Royalty.test.js
Comment (JulissaDantes) [reply test/token/ERC721/extensions/ERC721Royalty.test.js:30]: Without this it doesnt pass the coverge test, it doesnt recognize it being tests on the `shouldSupportInterfaces ` method

Comment (frangio) [reply test/token/ERC721/extensions/ERC721Royalty.test.js:30]: In `shouldSupportIntrefaces` we skip this check in coverage because it can fail due to gas limit. (Coverage instrumentation increases gas cost and ERC165 has a 30k gas requirement.)

So it is acceptable if coverage doesn't recognize it.

https://github.com/OpenZeppelin/openzeppelin-contracts/blob/d95cdaae8e26b3e479ca2ad70cc7c5a303444392/test/utils/introspection/SupportsInterface.behavior.js#L103-L104

Comment (JulissaDantes) [reply test/token/ERC721/extensions/ERC721Royalty.test.js]: commit :3d826f8f

File: contracts/token/ERC1155/extensions/ERC1155Royalty.sol
Comment (frangio) [inline contracts/token/ERC1155/extensions/ERC1155Royalty.sol:28]: Since `ERC2981` implements the interface, this should be defined in that contract.

File: contracts/token/ERC1155/extensions/ERC1155Royalty.sol
Comment (Amxx) [inline contracts/token/ERC1155/extensions/ERC1155Royalty.sol:26]: this is already in ERC2981 (duplicate)

File: contracts/token/ERC721/extensions/ERC721Royalty.sol
Comment (Amxx) [inline contracts/token/ERC721/extensions/ERC721Royalty.sol:26]: this is already in ERC2981 (duplicate)

File: contracts/token/ERC1155/extensions/ERC1155Royalty.sol
Comment (JulissaDantes) [reply contracts/token/ERC1155/extensions/ERC1155Royalty.sol:26]: Since both ERC1155 and ERC2981 inherit from ERC165 it needs this implementation or it wont compile

File: contracts/token/ERC721/extensions/ERC721Royalty.sol
Comment (JulissaDantes) [reply contracts/token/ERC721/extensions/ERC721Royalty.sol:26]: Since both ERC721 and ERC2981 inherit from ERC165 it needs this implementation or it wont compile

File: contracts/token/ERC1155/extensions/ERC1155Royalty.sol
Comment (sreyemnayr) [inline contracts/token/ERC1155/extensions/ERC1155Royalty.sol:40]: I don't understand why burning some number of tokens in an ERC-1155 should trigger a royalties reset. It makes sense for ERC791, but not for ERC1155, unless I'm missing something.

File: contracts/token/ERC1155/extensions/ERC1155Royalty.sol
Comment (Amxx) [reply contracts/token/ERC1155/extensions/ERC1155Royalty.sol:26]: at most you need 
```
return super.supportsInterface(interfaceId);
```

you clearly don't need the `interfaceId == type(IERC2981).interfaceId || ` part

File: contracts/token/ERC1155/extensions/ERC1155Royalty.sol
Comment (Amxx) [reply contracts/token/ERC1155/extensions/ERC1155Royalty.sol:40]: right, this should not happen unless ALL token (for this id) are burned.

File: contracts/token/ERC1155/extensions/ERC1155Royalty.sol
Comment (JulissaDantes) [reply contracts/token/ERC1155/extensions/ERC1155Royalty.sol:26]: This was modified on `7ab210f`

File: contracts/token/ERC721/extensions/ERC721Royalty.sol
Comment (JulissaDantes) [reply contracts/token/ERC721/extensions/ERC721Royalty.sol:26]: This was modified on `7ab210f`

File: contracts/token/ERC1155/extensions/ERC1155Royalty.sol
Comment (JulissaDantes) [reply contracts/token/ERC1155/extensions/ERC1155Royalty.sol:41]: Solved on `c2c11bb`

**Review by turbowars — COMMENTED** [review]
Must commend this is some deep work.

**Review by frangio — APPROVED** [review]

**Comment by frangio** [comment]
:rocket: Thanks everyone! This will be released in 4.5 later this month.

**Comment by hasandogu** [comment]
This is exciting, thank you! Any chance there could be an intermediate release? We are in the middle of implementing our own EIP2981 but would much rather use this.

**Comment by frangio** [comment]
@hasandogu Sorry, can't release this immediately, it has to go through our release process for review and bug bounty.

We're targetting a release candidate next week though, you're welcome to use it!

**Comment by hasandogu** [comment]
> We're targetting a release candidate next week though, you're welcome to use it!

Thank you @frangio. We'll wait for the release candidate and see if we can contribute to the bug bounty.

**Comment by trivikramak** [comment]
Hi @frangio, waiting for this release! Do we have a release candidate?
Also, any updated timeline for 4.5? and is this confirmed to be included in 4.5?

**Pull request merged**
